sherpa/bin/offline_server.py

```diff
@@ -0,0 +1,388 @@
+        """This function extracts features from the feature_queue,
+        batches them up, sends them to the RNN-T model for computation
+        and decoding.
+        """
+        while True:
+            if self.feature_queue.empty():
+                await asyncio.sleep(self.max_wait_ms / 1000)
+                continue
+            batch = []
+            try:
+                while len(batch) < self.batch_size:
+                    item = self.feature_queue.get_nowait()
+                    batch.append(item)
+            except asyncio.QueueEmpty:
+                pass
+
+            feature_list = [b[0] for b in batch]
+
+            loop = asyncio.get_running_loop()
+            self.counter = (self.counter + 1) % len(self.nn_models)
+            model = self.nn_models[self.counter]
+
+            hyp_tokens = await loop.run_in_executor(
+                self.nn_pool,
+                run_model_and_do_greedy_search,
+                model,
+                feature_list,
+            )
+
+            for i, hyp in enumerate(hyp_tokens):
+                self.feature_queue.task_done()
+                future = batch[i][1]
+                loop.call_soon(future.set_result, hyp)
+
+    async def compute_features(self, samples: torch.Tensor) -> torch.Tensor:
+        """Compute the fbank features for the given audio samples.
+
+        Args:
+          samples:
+            A 1-D torch.float32 tensor containing the audio samples. Its
+            sampling rate should be the one as expected by the feature
+            extractor. Also, its range should match the one used in the
+            training.
+        Returns:
+          Return a 2-D tensor of shape (num_frames, feature_dim) containing
+          the features.
+        """
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(
+            self.feature_extractor_pool,
+            self.feature_extractor,  # it releases the GIL
+            samples,
+        )
+
+    async def compute_and_decode(
+        self,
+        features: torch.Tensor,
+    ) -> List[int]:
+        """Run the RNN-T model on the features and do greedy search.
+
+        Args:
+          features:
+            A 2-D tensor of shape (num_frames, feature_dim).
+        Returns:
+          Return a list of token IDs containing the decoded results.
+        """
+        loop = asyncio.get_running_loop()
+        future = loop.create_future()
+        await self.feature_queue.put((features, future))
+        await future
+        return future.result()
+
+    async def handle_connection(
+        self,
+        socket: websockets.WebSocketServerProtocol,
+    ):
+        """Receive audio samples from the client, process it, and sends
+        deocoding result back to the client.
+
+        Args:
+          socket:
+            The socket for communicating with the client.
+        """
+        logging.info(f"Connected: {socket.remote_address}")
+        while True:
```

Comment:
This shows what each connection does:
 - receive audio samples from the client
 - compute features
 - feed the features to the queue, waiting for decoding
 - fetch decoding results and send them to the client